Add `label_transparent` property to EditorProperty

fire-forge opened the pull request:
This PR adds a new property to EditorProperty, `label_transparent`, which makes the label text display at 0.5 transparency. I added this property because it previously checked for a period in `label` to make text partially transparent (used for project settings feature tag overrides), but there was a comment in the code saying that this functionality should be moved to project settings instead. As an added bonus, it's exposed to GDScript so plugins can make use of it.

This comment here is the main reason for adding this:
```cpp
if (label.contains(".")) {
	// FIXME: Move this to the project settings editor, as this is only used
	// for project settings feature tag overrides.
	color.a = 0.5;
}
```

Alternatively, this could be done as a `label_transparency` property or a `label_modulate` theme property, but I'm not sure how much more useful that would really be.

fire-forge committed:
Add label_transparent property to EditorProperty

pycbouh commented:
Rounded corners seem nice, however...

> `label_centered` makes the label text and buttons centered vertically when `true` and top-aligned when `false`.
> * This is a replacement for `EditorProperty.set_label_reference()` which was used by several EditorProperty types to make their text top-aligned but wasn't exposed to GDScript.

I don't see a good reason to introduce a more limited replacement. Can you argue why this is better than exposing the method that already exists and is used?

> * Fixed a bug where label text could overlap children by a few pixels.
> * Fixed a bug where the revert icon would overlap text if it had a checkbox (mainly in Control theme overrides).

I think bug fixes should be PR'd separately from enhancement proposals. We can merge them now, without much discussion.

fire-forge commented:
I made the option more limited for a few reasons:
1. `set_label_reference` was only ever used to make text top-aligned, so `label_centered` essentially achieves the same thing.
2. It's easier to use and understand. 
3. Using the height and position of another node to determine the label's height and position feels like a hack.

I can't think of many cases where a plugin developer would want to position the label text somewhere other than the center or top, except maybe the bottom in which case I could change it to an enum like `Label.vertical_alignment`. Would this deserve making a proposal for more discussion?

I'll change the commit message and move the bug fixes into another PR.

Thanks for letting me know that I should keep bug fixes and improvements in separate PRs. I'll keep that in mind for any future contributions I do.

pycbouh commented:
> I can't think of many cases where a plugin developer would want to position the label text somewhere other than the center or top, except maybe the bottom in which case I could change it to an enum like `Label.vertical_alignment`. Would this deserve making a proposal for more discussion?

A proposal is always appreciated to have dedicated space for a discussion. Though plugins are always too niche to produce much of it.

> 3. Using the height and position of another node to determine the label's height and position feels like a hack.

Ideally, we should always either top-align or center-align. But it seems there is (or used to be at some point?) a need to align the label to a specific part of the property editor. It makes sense for more complex controls to have this option, though I can't think immediately of a single use for it.

At the very least, if we were to change this, an enum is definitely better than "centered" boolean. An enum would make the choice much clearer.

fire-forge commented:
I made a new PR for the bug fixes #58125. I'll remove them from this PR and add the changes suggested by @pycbouh.

fire-forge commented:
@pycbouh I removed the label alignment property because I wasn't what options should be given. I'll probably open a proposal so it can be discussed before opening a new PR with those options separately from this one.

I also updated the main post to reflect the current state of this PR.

pycbouh commented:
This looks lovely! Not sure from your description, where the highlight on the offset array property go? Or did you just not focus on it for the second screenshot?

fire-forge commented:
It still has the highlight, I guess I just forgot to select it when taking the second screenshot.

fire-forge commented:
Rebased and fixed an issue with properties that don't have a control on the right side having too much vertical padding.

fire-forge commented:
Rebased and added margins to EditorPropertyLocalizableString to match Array and Dictionary.

fire-forge commented:
I moved the rounded corners into a new PR, https://github.com/godotengine/godot/pull/59770, as this PR was doing too many things originally and I'm not sure that this even was the best way to solve this problem

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -268,15 +268,16 @@ void EditorProperty::_notification(int p_what) {
 			int ofs = get_theme_constant(SNAME("font_offset"));
 			int text_limit = text_size - ofs;
 
+			Color normal_color = Color(1, 1, 1);
+			Color hover_color = Color(1.2, 1.2, 1.2);
```
```suggestion
			const Color normal_color = Color(1, 1, 1);
			const Color hover_color = Color(1.2, 1.2, 1.2);
```

KoBeWi reviewed (dismissed):
Needs another rebase 🙃

AThousandShips commented on doc/classes/EditorProperty.xml:
```diff
@@ -82,6 +82,9 @@
 			Set this property to change the label (if you want to show one).
 		</member>
+		<member name="label_transparent" type="bool" setter="set_label_transparent" getter="is_label_transparent" default="false">
+			If [code]true[/code], the label text will show at 0.5 transparency. This is used for feature tag overrides in the Project Settings menu.
```
```suggestion
			If [code]true[/code], the label text will show at [code]0.5[/code] transparency. This is used for feature tag overrides in the Project Settings menu.
```
